### PR TITLE
[Profiler] Detect Single Step Instrumentation

### PR DESCRIPTION
**Pull request by chrisnas**
## Summary of changes
Detect Single Step Instrumentation deploument and activation

## Reason for change
When APM is deployed via SSI, some heuristics will be needed to avoid impacting applications performance. The detection of this deployment and activation mode should be available via the profiler configuration.

## Implementation details
Relies on DD_INJECTION_ENABLED environment variable presence and content (if it contains "profiling", it means that customer has explicitely enabled the profiler during the Agent installation)

## Test coverage
Corresponding native tests have been added.

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp [hunk 1]
```diff
@@ -604,3 +615,30 @@ bool Configuration::IsEnvironmentValueSet(shared::WSTRING const& name, T& value)
...
+    if (!shared::EnvironmentExist(EnvironmentVariables::SsiDeployed))
+    {
+        ssiDeployed = false;
+        ssiEnabled = false;
+        return;
+    }
+
+    ssiDeployed = true;
+
+    auto r = shared::Trim(shared::GetEnvironmentValue(EnvironmentVariables::SsiDeployed));
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:635]: no need to call `shared::Trim`.
`shared::GetEnrionmentVariableValue` calls it.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp [hunk 2]
```diff
@@ -604,3 +615,30 @@ bool Configuration::IsEnvironmentValueSet(shared::WSTRING const& name, T& value)
...
+    if (r.empty())
+    {
+        ssiEnabled = false;
+        return;
+    }
+
+    auto pos = r.find(WStr("profiling"),0);
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:642]: no need to specify the start position `(0)` . It's the default.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp [hunk 3]
```diff
@@ -604,3 +615,30 @@ bool Configuration::IsEnvironmentValueSet(shared::WSTRING const& name, T& value)
...
+    ssiEnabled = (pos != -1);
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:643]: prefer checking against `shared::WSTRING::npos` instead of `-1`

**Review by gleocadie — APPROVED** [review]
LGTM
Additional note: could we log this the `CorProfilerCallback` so in case of customer report we know how it was deployed ?

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp (hunk 2 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:642]: ✅ done

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp (hunk 3 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:643]: ✅ done

**Comment by chrisnas** [comment]
> LGTM Additional note: could we log this the `CorProfilerCallback` so in case of customer report we know how it was deployed ?

This PR is just for detection API: the next one will take care of the integration into the profiler with logging and telemetry

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5184)](https://github.com/DataDog/dd-trace-dotnet/tree/64aca5ff653ec50d2bdbd4128d7b6c72ad9ead9c)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/d3b7b308d25bc3aad2198a34ba163030107a9f75)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (71ms)  : 68, 74
     .   : milestone, 71,
    master - mean (76ms)  : 67, 85
     .   : milestone, 76,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (1,002ms)  : 812, 1192
     .   : milestone, 1002,
    master - mean (988ms)  : 965, 1011
     .   : milestone, 988,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (111ms)  : 106, 116
     .   : milestone, 111,
    master - mean (110ms)  : 107, 113
     .   : milestone, 110,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (703ms)  : 677, 730
     .   : milestone, 703,
    master - mean (714ms)  : 694, 734
     .   : milestone, 714,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (93ms)  : 90, 97
     .   : milestone, 93,
    master - mean (94ms)  : 91, 97
     .   : milestone, 94,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (657ms)  : 637, 677
     .   : milestone, 657,
    master - mean (663ms)  : 641, 686
     .   : milestone, 663,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (187ms)  : 180, 194
     .   : milestone, 187,
    master - mean (188ms)  : 184, 191
     .   : milestone, 188,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (1,036ms)  : 1006, 1065
     .   : milestone, 1036,
    master - mean (1,053ms)  : 1027, 1080
     .   : milestone, 1053,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (271ms)  : 267, 276
     .   : milestone, 271,
    master - mean (271ms)  : 265, 278
     .   : milestone, 271,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (1,004ms)  : 971, 1038
     .   : milestone, 1004,
    master - mean (1,035ms)  : 1001, 1068
     .   : milestone, 1035,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5184) - mean (260ms)  : 255, 265
     .   : milestone, 260,
    master - mean (262ms)  : 258, 266
     .   : milestone, 262,

    section CallTarget+Inlining+NGEN
    This PR (5184) - mean (977ms)  : 954, 1001
     .   : milestone, 977,
    master - mean (1,001ms)  : 978, 1024
     .   : milestone, 1001,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5184)](https://github.com/DataDog/dd-trace-dotnet/tree/68655d94779fc1872edbc2dfa546cb25b5f3fced)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/d3b7b308d25bc3aad2198a34ba163030107a9f75)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5184) (10.731M)   : 0, 10730845
    master (10.904M)   : 0, 10903785
    benchmarks/2.9.0 (11.116M)   : 0, 11116034

    section Automatic
    This PR (5184) (7.418M)   : 0, 7418207
    master (7.548M)   : 0, 7547783
    benchmarks/2.9.0 (7.891M)   : 0, 7891342

    section Trace stats
    This PR (5184) (7.635M)   : 0, 7634881
    master (7.902M)   : 0, 7902379

    section Manual
    This PR (5184) (9.494M)   : 0, 9493939
    master (9.632M)   : 0, 9631577

    section Manual + Automatic
    This PR (5184) (7.026M)   : 0, 7025599
    master (7.263M)   : 0, 7262759

    section Version Conflict
    This PR (5184) (6.394M)   : 0, 6394464
    master (6.604M)   : 0, 6603647

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5184) (9.586M)   : 0, 9585802
    master (9.355M)   : 0, 9354781
    benchmarks/2.9.0 (9.562M)   : 0, 9562342

    section Automatic
    This PR (5184) (6.638M)   : 0, 6637980
    master (6.646M)   : 0, 6646272

    section Trace stats
    This PR (5184) (7.042M)   : 0, 7041892
    master (6.782M)   : 0, 6782074

    section Manual
    This PR (5184) (8.286M)   : 0, 8286455
    master (8.426M)   : 0, 8425694

    section Manual + Automatic
    This PR (5184) (6.276M)   : 0, 6276000
    master (6.313M)   : 0, 6313053

    section Version Conflict
    This PR (5184) (5.681M)   : 0, 5680510
    master (5.623M)   : 0, 5622957

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5184) (10.256M)   : 0, 10256498
    master (10.039M)   : 0, 10039221
    benchmarks/2.9.0 (10.060M)   : 0, 10060320

    section Automatic
    This PR (5184) (7.121M)   : 0, 7120917
    master (7.158M)   : 0, 7157695
    benchmarks/2.9.0 (7.535M)   : 0, 7534759

    section Trace stats
    This PR (5184) (7.439M)   : 0, 7439249
    master (7.442M)   : 0, 7442467

    section Manual
    This PR (5184) (8.764M)   : 0, 8763786
    master (8.814M)   : 0, 8814383

    section Manual + Automatic
    This PR (5184) (6.853M)   : 0, 6853318
    master (6.903M)   : 0, 6903478

    section Version Conflict
    This PR (5184) (6.157M)   : 0, 6157327
    master (6.144M)   : 0, 6143514

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    master (7.398M)   : 0, 7398268
    benchmarks/2.9.0 (7.618M)   : 0, 7617642

    section No attack
    master (1.836M)   : 0, 1835561
    benchmarks/2.9.0 (3.121M)   : 0, 3120509

    section Attack
    master (1.457M)   : 0, 1457427
    benchmarks/2.9.0 (2.488M)   : 0, 2488303

    section Blocking
    master (3.175M)   : 0, 3174901

    section IAST default
    master (6.462M)   : 0, 6462305

    section IAST full
    master (5.843M)   : 0, 5843231

    section Base vuln
    master (0.991M)   : 0, 991340

    section IAST vuln
    master (0.892M)   : 0, 892185

```

**Pull request merged**